include/dlaf/communication/datatypes.h

```diff
@@ -10,25 +10,67 @@
 
 #pragma once
 
-#include "dlaf/communication/internal/datatypes.h"
+#include "dlaf/mpi_header.h"
+
+#include <complex>
 
 namespace dlaf {
 namespace comm {
 
-extern template MPI_Datatype mpi_datatype<char>::type;
-extern template MPI_Datatype mpi_datatype<short>::type;
-extern template MPI_Datatype mpi_datatype<int>::type;
-extern template MPI_Datatype mpi_datatype<long>::type;
-extern template MPI_Datatype mpi_datatype<long long>::type;
-extern template MPI_Datatype mpi_datatype<unsigned char>::type;
-extern template MPI_Datatype mpi_datatype<unsigned short>::type;
-extern template MPI_Datatype mpi_datatype<unsigned int>::type;
-extern template MPI_Datatype mpi_datatype<unsigned long>::type;
-extern template MPI_Datatype mpi_datatype<unsigned long long>::type;
-extern template MPI_Datatype mpi_datatype<float>::type;
-extern template MPI_Datatype mpi_datatype<double>::type;
-extern template MPI_Datatype mpi_datatype<bool>::type;
-extern template MPI_Datatype mpi_datatype<std::complex<float>>::type;
-extern template MPI_Datatype mpi_datatype<std::complex<double>>::type;
+/// @brief mapper between language types and basic MPI_Datatype
+template <typename T>
+struct mpi_datatype {
+  static MPI_Datatype type;
+};
+
+/// helper for mapping also const types
+template <typename T>
+struct mpi_datatype<const T> : public mpi_datatype<T> {};
+
+template <>
+MPI_Datatype mpi_datatype<char>::type = MPI_CHAR;
+
+template <>
+MPI_Datatype mpi_datatype<short>::type = MPI_SHORT;
+
+template <>
+MPI_Datatype mpi_datatype<int>::type = MPI_INT;
+
+template <>
+MPI_Datatype mpi_datatype<long>::type = MPI_LONG;
+
+template <>
+MPI_Datatype mpi_datatype<long long>::type = MPI_LONG_LONG;
+
+template <>
+MPI_Datatype mpi_datatype<unsigned char>::type = MPI_UNSIGNED_CHAR;
+
+template <>
+MPI_Datatype mpi_datatype<unsigned short>::type = MPI_UNSIGNED_SHORT;
+
+template <>
+MPI_Datatype mpi_datatype<unsigned int>::type = MPI_UNSIGNED;
+
+template <>
+MPI_Datatype mpi_datatype<unsigned long>::type = MPI_UNSIGNED_LONG;
+
+template <>
+MPI_Datatype mpi_datatype<unsigned long long>::type = MPI_UNSIGNED_LONG_LONG;
+
+template <>
+MPI_Datatype mpi_datatype<float>::type = MPI_FLOAT;
+
+template <>
+MPI_Datatype mpi_datatype<double>::type = MPI_DOUBLE;
+
+template <>
+MPI_Datatype mpi_datatype<bool>::type = MPI_CXX_BOOL;
+
+template <>
+MPI_Datatype mpi_datatype<std::complex<float>>::type = MPI_CXX_FLOAT_COMPLEX;
+
+template <>
+MPI_Datatype mpi_datatype<std::complex<double>>::type = MPI_CXX_DOUBLE_COMPLEX;
```

Comment:
This is the same as in the original PR : https://github.com/eth-cscs/DLA-Future/pull/170 . There are issues with it related to multiple definitions error when static class variables are defined in headers and included in multiple translation units that are linked

Reply:
Right! I forgot it! Thank you ;)